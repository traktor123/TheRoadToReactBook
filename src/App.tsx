import * as React from 'react';
import axios from 'axios';

import styles from './App.module.css';
import './App1.scss';
import './App2.css';
import clsx from 'clsx';

import styled from 'styled-components';

import { ReactComponent as Check } from './assets/check.svg';

import { FaBeer, FaApple, FaFacebook } from "react-icons/fa";

/*CSS in JS Styled Components*/
const StyledContainer = styled.div`
height: 100vw;
padding: 20px;
background: #83a4d4;
background: linear-gradient(to left, #b6fbff, #83a4d4);
color: #171212;
`;

const StyledHeadlinePrimary = styled.h1`
font-size: 48px;
font-weight: 300;
letter-spacing: 2px;
color: blue;
`;

const StyledItem = styled.li`
display: flex;
align-items: center;
padding-bottom: 5px;
`;

type StyledColumnProps = {
  width: string;
}

const StyledColumn = styled.span<StyledColumnProps>`
padding: 0 5px;
white-space: nowrap;
overflow: hidden;
white-space: nowrap;
text-overflow: ellipsis;
a {
color: inherit;
}
width: ${(props) => props.width};
`;

const StyledButton = styled.button`
  background: transparent;
  border: 1px solid #171212;
  padding: 5px;
  cursor: pointer;
  transition: all 0.1s ease-in;

  &:hover {
    background: #171212;
    color: #ffffff;
  }
`;

const StyledButtonSmall = styled(StyledButton)`
  padding: 5px;
`;

const StyledButtonLarge = styled(StyledButton)`
  padding: 10px;
`;

const StyledSearchForm = styled.form`
  padding: 10px 0 20px 0;
  display: flex;
  align-items: baseline;
`;

const StyledLabel = styled.label`
  border-top: 1px solid #171212;
  border-left: 1px solid #171212;
  padding-left: 5px;
  font-size: 24px;
`;

const StyledInput = styled.input`
  border: none;
  border-bottom: 1px solid #171212;
  background-color: transparent;
  font-size: 24px;
`;


/*Data Types*/

type Story = {
  title: string,
  url: string,
  author: string,
  num_comments: number,
  points: number,
  objectID: number
}

/*Component Types*/

type ListProps = {
  list: Story[];
  onRemoveItem: (item: Story) => void;
};

type ItemProps = {
  /*
  title: string,
  url: string,
  author: string,
  num_comments: number,
  points: number,
  objectID: number,
  */
  item: Story;
  onRemoveItem: (item: Story) => void;
}

type InputWithLabelProps = {
  onInputChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  value?: string;
  id: string;
  type?: string;
  children?: React.ReactNode;
  isFocused?: boolean;
};

/*Reducers*/
type StoriesState = {
  data: Story[],
  isLoading: boolean;
  isError: boolean;
}

type StoriesFetchInit = {
  type: 'STORIES_FETCH_INIT';
}

type StoriesFetchInit1 = {
  type: 'STORIES_FETCH_INIT1';
}

type StoriesFetchSuccessAction = {
  type: 'STORIES_FETCH_SUCCESS';
  payload: Story[];
};

type StoriesFetchFailureAction = {
  type: 'STORIES_FETCH_FAILURE';
};

type StoriesRemoveAction = {
  type: 'REMOVE_STORY';
  payload: Story;
};

type StoriesAction = StoriesFetchInit | StoriesFetchSuccessAction | StoriesFetchFailureAction | StoriesRemoveAction | StoriesFetchInit1;

const storiesReducer = (
  state: StoriesState,
  action: StoriesAction
) => {
  switch (action.type) {
    case 'STORIES_FETCH_INIT':
      return {
        ...state,
        isLoading: true,
        isError: false,
      };
    case 'STORIES_FETCH_INIT1':
      return {
        ...state,
        isLoading: false,
        isError: false,
      };
    case 'STORIES_FETCH_SUCCESS':
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload,
      };
    case 'STORIES_FETCH_FAILURE':
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    case 'REMOVE_STORY':
      return {
        ...state,
        data: state.data.filter(
          (story) => action.payload.objectID !== story.objectID
        ),
      };
    default:
      throw new Error();
  }
};

// Data
const API_ENDPOINT = 'https://hn.algolia.com/api/v1/search?query=';

/* functions */

const getAsyncStories = async (url: string = '') => {
  console.log(url);
  let result = await axios.get(url)
  return result.data.hits;

  /*
  return fetch(url) // B
    .then((response) => response.json())
    .then(data => data.hits.map((hit: any) => ({
      title: hit.title,
      url: hit.url,
      author: hit.author,
      num_comments: hit.num_comments,
      points: hit.points,
      objectID: hit.objectID,
    })))
  */

  /*
  return new Promise<{ data: { stories: Story[] } }>((resolve, reject) => {
    setTimeout(() => {
      //throw "Error while loading Stories from remote Repository."; //will not work because within timeout!
      // reject("Error while loading Stories from remote Repository.");
      resolve({ data: { stories: initialStoriesList } })
    }, 3000);
  });
  */
}

const useStorageState = (key: string, initialState: string): [string, (initialState: string) => void] =>
//[string, React.Dispatch<React.SetStateAction<string>>]   
{
  const isMounted = React.useRef(false);

  const [value, setValue] = React.useState(localStorage.getItem(key) || initialState);

  React.useEffect(() => {
    if (!isMounted.current) {
      isMounted.current = true;
    } else {
      console.log('A');
      localStorage.setItem(key, value);
    }
  }, [value, key]);

  return [value, setValue];
};

/* Components */

const getSumComments = (stories: { data: Story[] }) => {
  console.log('C');
  return stories.data.reduce(
    (result, value) => result + value.num_comments,
    0
  );
};

export const App = () => {
  const [searchTerm, setSearchTerm] = useStorageState('searchTerm', 'React');
  const [stories, dispatchStories] = React.useReducer(storiesReducer, { data: [], isLoading: false, isError: false });
  const [url, setUrl] = React.useState(`${API_ENDPOINT}${searchTerm}`);

  const handleFetchStories = React.useCallback(async () => {
    dispatchStories({ type: 'STORIES_FETCH_INIT' });
    dispatchStories({ type: 'STORIES_FETCH_INIT1' });
    dispatchStories({ type: 'STORIES_FETCH_INIT' });
    try {
      const result = await getAsyncStories(url);
      dispatchStories({
        type: 'STORIES_FETCH_SUCCESS',
        payload: result/*.data.stories*/, //D
      });
    }
    catch (error) {
      dispatchStories({ type: 'STORIES_FETCH_FAILURE' });
    }
  }, [url]);

  React.useEffect(() => {
    console.log('How many times do I log?');
    handleFetchStories();
  }, [handleFetchStories]);

  const handleRemoveStory = React.useCallback((item: Story) => {
    dispatchStories({
      type: 'REMOVE_STORY',
      payload: item,
    });
  }, []);

  /*
  let searchedStories = searchTerm === ''
    ? stories.data
    : stories.data.filter(s => s.title.toLowerCase().includes(searchTerm?.toLowerCase() ?? ""));
  */
  let searchedStories = stories.data;

  const handleSearchInput = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(event.target.value);
  };

  const handleSearchSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    setUrl(`${API_ENDPOINT}${searchTerm}`);
    event.preventDefault();
  };

  console.log('B:App');

  //const sumComments = getSumComments(stories);
  const sumComments = React.useMemo(
    () => getSumComments(stories),
    [stories]
  );  

  return (
    // <div className={styles.container}>
    <StyledContainer>
      Icons from react-icons/fa:<span><FaBeer /><FaApple /><FaFacebook /></span>
      {/* <h1 className={styles.headlinePrimary}>My Hacker Stories</h1> */}
      {/* <StyledHeadlinePrimary>My Hacker Stories</StyledHeadlinePrimary> */}
      <h1>My Hacker Stories with {sumComments} comments.</h1>
      <SearchForm onSearchSubmit={handleSearchSubmit} onSearchInput={handleSearchInput} searchTerm={searchTerm} />
      {/* <hr /> */}
      {stories.isError && <p style={{ color: 'red' }}>Something went wrong ...</p>}
      {
        stories.isLoading
          ? (<strong><h1>Loading...</h1></strong>)
          : (<List list={searchedStories} onRemoveItem={handleRemoveStory} />)
      }
      {/* </div> */}
    </StyledContainer>
  );
};

type SearchFormProps = {
  onSearchSubmit?: (event: React.FormEvent<HTMLFormElement>) => void;
  onSearchInput: (event: React.ChangeEvent<HTMLInputElement>) => void;
  searchTerm: string;
}

const SearchForm: React.FC<SearchFormProps> = ({ onSearchSubmit: onSearchSubmit, onSearchInput: onSearchInput, searchTerm }) => {
  return (
    <StyledSearchForm onSubmit={onSearchSubmit}>
      <InputWithLabel
        onInputChange={onSearchInput}
        value={searchTerm}
        id="search" isFocused={true}>
        <strong>Search:</strong>
      </InputWithLabel>
      &nbsp;
      {/* <button type="submit" disabled={!searchTerm} className={`${styles.button} ${styles.buttonSmall}`}>Submit</button> */}
      {/* <button type="submit" disabled={!searchTerm} className="smallButton1">Submit</button> */}
      <StyledButtonLarge type="submit" disabled={!searchTerm}>Submit</StyledButtonLarge>
    </StyledSearchForm>)
}

/*
const InputWithLabel: React.FC<InputWithLabelProps> = ({ value, onInputChange, id, type = 'text', children, isFocused = false }) => {
  const inputRef = React.useRef<HTMLInputElement>(null);
  React.useEffect(() => {
    if (isFocused && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isFocused]);

  return (
    <div>
      <label htmlFor={id}>{children}</label>
      &nbsp;
      <input
        id={id}
        type={type}
        onChange={onInputChange}
        value={value}
        ref={inputRef} />
      <p>
        Searching for <strong>{value}</strong>
      </p>
    </div>)
}
*/

/*----------Class component examples----------*/

type App1Props = {
  searchTerm: string;
}

type App1State = {
  searchTerm: string;
}

export class App1 extends React.Component<App1Props, App1State> {
  constructor(props: App1Props) {
    super(props);
    this.state = {
      searchTerm: 'React1',
    };
  }
  render() {
    const { searchTerm } = this.state;
    return (
      <div>
        <h1>My Hacker Stories from Class compoent</h1>
        <SearchForm
          searchTerm={searchTerm}
          onSearchInput={(event) => this.setState({ searchTerm: event.target.value })}
        />
      </div>
    );
  }
}

class InputWithLabel extends React.Component<InputWithLabelProps> {

  private inputRef: React.RefObject<HTMLInputElement>;

  constructor(props: InputWithLabelProps) {
    super(props);
    this.inputRef = React.createRef();
  }

  //Class component’s lifecycle method 
  componentDidMount() {
    /*
    if (this.props.isFocused) {
      this.inputRef.current?.focus();
    }
    */
  }

  render() {
    const {
      id,
      value,
      type = 'text',
      onInputChange,
      children,
    } = this.props;
    return (
      <>
        <StyledLabel htmlFor={id}>{children}</StyledLabel>
        &nbsp;
        <StyledInput
          ref={this.inputRef}
          id={id}
          type={type}
          value={value}
          onChange={onInputChange} />
      </>
    );
  }
}

/*----------End class component examples----------*/

const List: React.FC<ListProps> = React.memo(({ list, onRemoveItem }) => {
  console.log('B:List');
  return <ul>
    {
      list.map((item) => (
        <div key={item.objectID}>
          <Item item={item} onRemoveItem={onRemoveItem} />
        </div>
      ))
    }
  </ul>
});

const isSmall: boolean = true;

const Item: React.FC<ItemProps> = ({
  item,
  onRemoveItem }) =>
(
  <StyledItem>
    <StyledColumn width="40%">
      <a href={item.url}>{item.title}</a>
    </StyledColumn>
    &nbsp;
    <StyledColumn width="30%">{item.author}</StyledColumn>
    &nbsp;
    <StyledColumn width="10%">{item.num_comments}</StyledColumn>
    &nbsp;
    <StyledColumn width="10%">{item.points}</StyledColumn>
    &nbsp;
    <StyledColumn width="10%">
      <button className={clsx(styles.button, { [styles.buttonSmall]: isSmall })}
        type="button"
        onClick={onRemoveItem.bind(null, item)}>
        <Check height="18px" width="18px" />
      </button>
    </StyledColumn>
  </StyledItem>
)

export default App;