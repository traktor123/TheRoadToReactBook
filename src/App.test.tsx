import { describe, it, expect, vi } from 'vitest';

import App, {
    storiesReducer,
    Item,
    SearchForm,
    StoriesAction,
    SearchFormProps
} from './App';

import {
    render,
    screen,
    fireEvent,
    waitFor,
} from '@testing-library/react';

import axios from 'axios';
vi.mock('axios');

describe('something truthy and falsy', () => {
    it('true to be true', () => {
        expect(true).toBeTruthy();
    });
    it('false to be false', () => {
        expect(false).toBeFalsy();
    });
});

it('true to be true', () => {
    expect(true).toBe(true);
});
it('false to be false', () => {
    expect(false).toBe(false);
});

describe('App component', () => {
    it('removes an item when clicking the Dismiss button', () => {
    });
    it('requests some initial stories from an API', () => {
    });
});

const storyOne = {
    title: 'React',
    url: 'https://reactjs.org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
};
const storyTwo = {
    title: 'Redux',
    url: 'https://redux.js.org/',
    author: 'Dan Abramov, Andrew Clark',
    num_comments: 2,
    points: 5,
    objectID: 1,
};
const stories = [storyOne, storyTwo];

describe('storiesReducer', () => {
    it('removes a story from all stories', () => {
        let action: StoriesAction = { type: "REMOVE_STORY", payload: storyTwo };
        let state = { data: stories, isLoading: false, isError: false }

        let newState = storiesReducer(state, action);

        let expectedSttate = {
            data: [storyOne],
            isLoading: false,
            isError: false
        };

        expect(newState).toStrictEqual(expectedSttate);
    });
});

describe('Item', () => {
    it('renders all properties', () => {
        render(<Item item={storyOne} />);
        //screen.debug();
        expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
        expect(screen.getByText('React')).toHaveAttribute(
            'href',
            'https://reactjs.org/'
        );
    });

    it('renders a clickable dismiss button', () => {
        render(<Item item={storyOne} />);
        //screen.getByRole('');
        expect(screen.getByRole('button')).toBeInTheDocument();
    });

    it('clicking the dismiss button calls the callback handler', () => {
        const handleRemoveItem = vi.fn();
        render(<Item item={storyOne} onRemoveItem={handleRemoveItem} />);
        fireEvent.click(screen.getByRole('button'));
        expect(handleRemoveItem).toHaveBeenCalledTimes(1);
    });

    describe('SearchForm', () => {
        const searchFormProps = {
            searchTerm: 'React',
            onSearchInput: vi.fn(),
            onSearchSubmit: vi.fn(),
        };
        it('renders the input field with its value', () => {
            render(<SearchForm {...searchFormProps} />);
            //screen.debug();
            expect(screen.getByDisplayValue('React')).toBeInTheDocument();
            //expect(screen.getByText('React')).toBeInTheDocument();
            expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
        });
        it('calls onSearchInput on input field change', () => {
            render(<SearchForm {...searchFormProps} />);
            fireEvent.change(screen.getByDisplayValue('React'), {
                target: { value: 'Redux' },
            });
            expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1);
        });
        it('calls onSearchSubmit on button submit click', () => {
            render(<SearchForm {...searchFormProps} />);
            fireEvent.submit(screen.getByRole('button'));
            expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
        });
    });

    describe('App', () => {
        it('succeeds fetching data', async () => {
            const promise = Promise.resolve({
                data: {
                    hits: stories,
                },
            });
            (axios.get as jest.Mock).mockImplementationOnce(() => promise);
            render(<App />);
            screen.debug();
            expect(screen.queryByText(/Loading/)).toBeInTheDocument();
            await waitFor(async () => {
                await promise;
                screen.debug();
                expect(screen.queryByText(/Loading/)).toBeNull();
                expect(screen.getByText('React')).toBeInTheDocument();
                expect(screen.getByText('Redux')).toBeInTheDocument();
                //expect(screen.getAllByRole('listitem').length).toBe(2);
                expect(screen.getAllByText('Dismiss').length).toBe(2);
                //console.log(data);
            });
        });
    });

    it('fails fetching data', async () => {
        const promise = Promise.reject();
        (axios.get as jest.Mock).mockImplementationOnce(() => promise);
        render(<App />);
        expect(screen.getByText(/Loading/)).toBeInTheDocument();
        try {
            await waitFor(async () => await promise);
        } catch (error) {
            expect(screen.queryByText(/Loading/)).toBeNull();
            expect(screen.queryByText(/went wrong/)).toBeInTheDocument();
        }
    });

    it('removes a story', async () => {
        const promise = Promise.resolve({
            data: {
                hits: stories,
            },
        });
        (axios.get as jest.Mock).mockImplementationOnce(() => promise);
        render(<App />);
        await waitFor(async () => await promise);
        await promise; //That should not be here
        expect(screen.getAllByText('Dismiss').length).toBe(2);
        expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
        fireEvent.click(screen.getAllByText('Dismiss')[0]);
        expect(screen.getAllByText('Dismiss').length).toBe(1);
        expect(screen.queryByText('Jordan Walke')).toBeNull();
    });

    it('searches for specific stories', async () => {
        const reactPromise = Promise.resolve({
            data: {
                hits: stories,
            },
        });
        const anotherStory = {
            title: 'JavaScript',
            url: 'https://en.wikipedia.org/wiki/JavaScript',
            author: 'Brendan Eich',
            num_comments: 15,
            points: 10,
            objectID: 3,
        };
        const javascriptPromise = Promise.resolve({
            data: {
                hits: [anotherStory],
            },
        });
        (axios.get as jest.Mock).mockImplementation((url: string) => {
            if (url.includes('React')) {
                return reactPromise;
            }
            if (url.includes('JavaScript')) {
                return javascriptPromise;
            }
            throw Error();
        });

        // Initial Render
        render(<App />);

        // First Data Fetching
        await waitFor(async () => await reactPromise);
        await reactPromise; //That should not be here

        expect(screen.queryByDisplayValue('React')).toBeInTheDocument();
        expect(screen.queryByDisplayValue('JavaScript')).toBeNull();
        expect(screen.queryByText('Jordan Walke')).toBeInTheDocument();
        expect(screen.queryByText('Dan Abramov, Andrew Clark')).toBeInTheDocument();

        expect(screen.queryByText('Brendan Eich')).toBeNull();

        // User Interaction -> Search
        fireEvent.change(screen.queryByDisplayValue('React'), {
            target: {
                value: 'JavaScript',
            },
        });
        expect(screen.queryByDisplayValue('React')).toBeNull();
        expect(screen.queryByDisplayValue('JavaScript')).toBeInTheDocument();

        fireEvent.submit(screen.queryByText('Submit'));
        // Second Data Fetching
        await waitFor(async () => await javascriptPromise);
        await javascriptPromise; //That should not be here
        expect(screen.queryByText('Jordan Walke')).toBeNull();
        expect(screen.queryByText('Dan Abramov, Andrew Clark')).toBeNull();
        expect(screen.queryByText('Brendan Eich')).toBeInTheDocument();
    });

    it('renders snapshot', () => {
        const searchFormProps: SearchFormProps = {
            onSearchSubmit: () => {},
            onSearchInput: () => {},
            searchTerm: "React"
        }

        const { container } = render(<SearchForm {...searchFormProps} />);
        expect(container.firstChild).toMatchSnapshot();
    });
});