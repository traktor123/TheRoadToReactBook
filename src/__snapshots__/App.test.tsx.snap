// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Item > renders snapshot 1`] = `
<form
  class="sc-kAycRU DMOuB"
>
  <label
    class="sc-imWZod kRtAqO"
    for="search"
  >
    <strong>
      <strong>
        Search:
      </strong>
    </strong>
  </label>
   
  <input
    class="sc-jXbVAB kTmaiu"
    id="search"
    type="text"
    value="React"
  />
   
  <button
    class="sc-dcJtft sc-gsFSjX gQElQm fhFcgD"
    type="submit"
  >
    Submit
  </button>
</form>
`;
